### PR TITLE
Fix 9265: Axis names overlap with labels

### PR DESCRIPTION
**Pull request by robin-gerling**
Follow-up of https://github.com/apache/echarts/pull/16825 and https://github.com/apache/echarts/pull/12236

## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [ ] new feature
- [ ] others

### What does this PR do?

This PR shrinks the grid size such that the axis labels and names are completely contained (Option (B) of https://github.com/apache/echarts/issues/9265#issuecomment-434022070).

### Fixed issues

- #9265: xAxis.nameGap and yAxis.nameGap should be set automatically given grid.containLabel

### Before: What was the problem?

When `grid.containLabel` is set to true, the axis names overlap with the labels, when there is no magical number for `nameGap` specified. Furthermore, the names may be not visible when specifying the `nameGap`, because the names get pushed out of the grid. Therefore, also the distance between the grid and the container needs to be adjusted.

before

### After: How does it behave after the fixing?

When `grid.containLabel` is set to true, the axis names do not overlap with labels and the grid size is shrinked such that the names are completely visible without the need to specify the grid to container distance or `nameGap`.

After

## Document Info

One of the following should be checked.

- [ ] This PR doesn't relate to document changes
- [ ] The document should be updated later
- [x] The document changes have been made in apache/echarts-doc#380

### ZRender Changes

- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

N.A.

### Merging options

- [ ] Please squash the commits into a single one when merging.

### CODE REVIEW

**Comment by DanielBogenrieder** [comment]
Hey @Ovilia, @100pah,

is there any chance we could get feedback on this PR?
We are happy to improve the PR if there is something wrong/missing, but it would be nice to get this into a release as soon as possible.

Greetings,
Daniel

**Comment by admsev** [comment]
@gooroodev review

**Comment by gooroodev** [comment]
### Summary of Changes

1. **AxisBuilder.ts**:
   - Imported `CartesianAxisModel`.
   - Modified the calculation of `gap` in the `axisName` builder to include `axisToNameGapStartGap` from `CartesianAxisModel`.

2. **axisHelper.ts**:
   - Added `computeNameBoundingRect` function to calculate the bounding rectangle for the axis name.
   - Added `isNameLocationCenter`, `isNameLocationStart`, and `isNameLocationEnd` helper functions.
   - Added types `CartesianAxisPositionMargins` and `ReservedSpace`.
   - Added `computeReservedSpace` function to calculate the reserved space for axis labels and names.

3. **AxisModel.ts**:
   - Added `inverseCartesianAxisPositionMap` for mapping axis positions.
   - Added `axisToNameGapStartGap` property to `CartesianAxisModel`.

4. **Grid.ts**:
   - Imported additional functions and types from `axisHelper`.
   - Modified the logic to compute the reserved space for axis labels and names.
   - Adjusted the grid rectangle dimensions based on the reserved space.

5. **axis-containLabelAndName.html**:
   - Added a new test HTML file to ensure axis labels and names do not overlap.
   - Included various test cases with different configurations for axis names and labels.

### Issues, Bugs, and Typos

1. **Unused Imports**:
   - In `Grid.ts`, `reduce` is imported but not used.

**Proposed Improvement**:
```diff
-import {isObject, each, indexOf, retrieve3, keys, reduce, map} from 'zrender/src/core/util';
+import {isObject, each, indexOf, retrieve3, keys, map} from 'zrender/src/core/util';
```

2. **Redundant Code**:
   - The function `isNameLocationCenter` is defined twice, once in `AxisBuilder.ts` and again in `axisHelper.ts`.

**Proposed Improvement**:
```diff
-function isNameLocationCenter(nameLocation: string) {
-    return nameLocation === 'middle' || nameLocation === 'center';
-}
```

### General Review of Code Quality and Style

- **Code Organization**: The code is well-organized, with clear separation of concerns between different modules. Functions and types are logically grouped.
- **Naming Conventions**: The naming conventions are consistent and descriptive, making it easy to understand the purpose of each function and variable.
- **Documentation**: The code includes comments and JSDoc annotations, which enhance readability and maintainability.
- **Test Coverage**: The addition of `axis-containLabelAndName.html` demonstrates a thorough approach to testing different configurations, ensuring robustness.
- **Style**: The code follows a consistent style, with proper indentation and spacing, making it easy to read.

Overall, the pull request addresses the issue effectively and maintains high code quality. The proposed improvements are minor and would enhance the clarity and efficiency of the code.

Yours, [Gooroo.dev](https://gooroo.dev). To receive reviews automatically, [install Github App](https://github.com/apps/gooroo-dev)

**Review by Ovilia — CHANGES_REQUESTED** [review]
Sorry for the late reply and thanks for your contribution!

I feel the changes are somewhat too complicated. Could we add the following code to the end of `axisHelper.ts#estimateLabelUnionRect`?

```ts
...
    const axisName = axisModel.get('name');
    if (axisName) {
        const nameLocation = axisModel.get('nameLocation');
        const nameGap = axisModel.get('nameGap') || 0;
        const nameRotate = axisModel.get('nameRotate') || 0;
        const nameTruncate = axisModel.get('nameTruncate');
        const textStyle = axisModel.getModel('nameTextStyle');

        const unrotatedSingleRect = textStyle.getTextRect(axisName);
        if (nameTruncate.maxWidth) {
            unrotatedSingleRect.width = Math.min(unrotatedSingleRect.width, nameTruncate.maxWidth);
        }

        const singleRect = rotateTextRect(unrotatedSingleRect, nameRotate);
        rect.union(singleRect);

        // TODO: also consider nameLocation, nameGap and etc. to control the rect
    }

    return rect;
}
```

**Comment by robin-gerling** [comment]
Unfortunately, yes, those changes are pretty complex.
If I understand your proposed changes, they would result in less functionality. However, those would simplify the logic.
My current proposed solution considers the labels of other axes, too. Those need to be taken into account in case the name location is either start or end, because otherwise the labels would overlap with the name, as shown in the provided figure. On the left is the proposed solution, on the right is the result of a simpler computation within `estimateLabelUnionRect` which includes the computation of the name and the computation of the label.
Therefore, by combining name and label to a single rect, the problem would not be solved for nameLocations start and end.
LabelNameOverlap

**Comment by DanielBogenrieder** [comment]
Hey @Ovilia,

I just saw that you gave a thumbs up. What does that mean? Do you prefer the simpler not always working solution or should we go for the more complex solution?

Greetings and thanks,
Daniel

**Comment by DanielBogenrieder** [comment]
Hey @Ovilia,

sorry to ping you again, but it would be really really nice to get this into some new version soon.

Let me know how we can help.

Greetings,
Daniel

**Comment by DanielBogenrieder** [comment]
Hey,

any updates on this one?

Cheers,
Daniel

File: src/coord/axisHelper.ts [hunk 1]
```diff
@@ -340,6 +341,24 @@ export function estimateLabelUnionRect(axis: Axis) {
...
+ * @param axis
+ * @return Be null/undefined if no name.
+ */
+export function computeNameBoundingRect(axis: Axis2D): BoundingRect {
+    const axisModel = axis.model;
+    if (!axisModel.get('name')) {
+        return;
+    }
+    const axisLabelModel = axisModel.getModel('nameTextStyle');
+    const unRotatedNameBoundingRect = axisLabelModel.getTextRect(axisModel.getModel('name').option);
```
Comment (Ovilia) [inline src/coord/axisHelper.ts:354]: We don't access `.option` directly. Use `axisModel.get('name')` instead.

File: src/coord/axisHelper.ts [hunk 2]
```diff
@@ -340,6 +341,24 @@ export function estimateLabelUnionRect(axis: Axis) {
...
+    const defaultRotation = axis.isHorizontal() || !isNameLocationCenter(axisModel.get('nameLocation')) ? 0 : -90;
+    const rotatedNameBoundingRect = rotateTextRect(
+        unRotatedNameBoundingRect, axisModel.getModel('nameRotate').option ?? defaultRotation
```
Comment (Ovilia) [inline src/coord/axisHelper.ts:357]: `axisModel.get('nameRotate')`

File: src/coord/axisHelper.ts [hunk 3]
```diff
@@ -399,3 +418,133 @@ export function unionAxisExtentFromData(dataExtent: number[], data: SeriesData,
...
+                    nameLocationIsStart && orthogonalAxisPositionIsLeft && nameRotationIsFirstOrThirdQuadrant
+                        || nameLocationIsStart && orthogonalAxisPositionIsRight && nameRotationIsSecondOrFourthQuadrant
+                        || nameLocationIsEnd && orthogonalAxisPositionIsLeft && nameRotationIsSecondOrFourthQuadrant
+                        || nameLocationIsEnd && orthogonalAxisPositionIsRight && nameRotationIsFirstOrThirdQuadrant
+                )
+            ) {
+                reservedSpace.name[namePositionOrthogonalAxis] = reservedNameSpace;
+            }
+        }
+    }
```
Comment (Ovilia) [inline src/coord/axisHelper.ts:548]: Can you explain the basic idea of these lines?

File: src/coord/cartesian/AxisModel.ts [hunk 4]
```diff
@@ -53,6 +60,12 @@ export class CartesianAxisModel extends ComponentModel<CartesianAxisOption>
 
     axis: Axis2D;
 
+    /**
+     * The gap between the axis and the name gap.
+     * Injected outside.
+     */
+    axisToNameGapStartGap: number = 0;
```
Comment (Ovilia) [inline src/coord/cartesian/AxisModel.ts:67]: Can you further explain what is the gap between the axis and the name gap?

**Review by Ovilia — CHANGES_REQUESTED** [review]
Thanks for your contribution and sorry for the late reply. I don't check the notification regularly because there are too many messages. Please pin me by email oviliazhang at gmail.com if I haven't reply for some time.

This PR has changed a lot so it may be challenging for me to understand. I need you to provide some extra explanation on some of the comments.

**Comment by Ovilia** [comment]
I've run with `npm run test:visual` and found that this PR may contain break changes like this:

image

I think the new result is better. It would be great to have this as default. Can we make this feature in the next major release 6.0.0, which may be released around 2025 Feb. If you want to use this feature earlier than this, you can still use the nightly build version under `next` tag after this PR is merged.

**Comment by Ovilia** [comment]
Can you check the diff with `test/custom-shape-morphing2.html`? 

image

It seems that the position of axis name and grid are changed even without overlapping. Can you explain why this break change happens?

**Comment by robin-gerling** [comment]
The `test/custom-shape-morphing2.html`looks different because the axis names are currently not included in the grid. With the new version, they are included and therefore decrease the size of the grid.
The green frame indicates the `gridRect` (with the default margins) before the calculations in `if (containLabel) {` (red frame afterwards). Therefore, with the current behaviour, the names would not be visible for a grid margin of e.g. 0.
CurrentBehaviour
NewBehaviour

**Review by Ovilia — APPROVED** [review]
Thanks for your contribution! This PR should be released in our next major release 6.0.0 since it contains break changes that may change the position of grid area.

**Comment by Ovilia** [comment]
Chinese document awaited, which I will add later.

**Pull request merged**